[native] Add S3 metrics reporting

xin-zhang2 opened the pull request:
## Description
<!---Describe your changes in detail-->

This PR introduced S3 file system metrics reporting in Prestissimo.

## Motivation and Context
<!---Why is this change required? What problem does it solve?-->
<!---If it fixes an open issue, please link to the issue here.-->
Presto Java provides PrestoS3FileSystemStats to collect and report the metrics related to S3 file system operations, and this PR is to fill the gap between Presto Java and Prestissimo. 

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [x] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.


```
== RELEASE NOTES ==
Prestissimo (native Execution) Changes
* Add runtime metrics collection for S3 Filesystem.

xin-zhang2 commented:
@majetideepak 
Could you help review this PR? Thanks!

majetideepak commented:
@xin-zhang2 can you change the commit title to `[native] Add S3 metrics reporting`. In Presto, the commit title is what gets reflected on master.

majetideepak commented:
@xin-zhang2 Can you also confirm you were able to test the S3 metrics locally? We do not have S3 test setup in CI.

xin-zhang2 commented:
@majetideepak. Thanks for the review. I have updated the commit title.


> @xin-zhang2 can you change the commit title to `[native] Add S3 metrics reporting`. In Presto, the commit title is what gets reflected on master.

xin-zhang2 commented:
@psnv03 
Could you help review this PR? Thanks.

ethanyzhang commented:
@majetideepak @xin-zhang2 do we need some release notes?
@steveburnett new doc?

majetideepak reviewed (changes requested):
@ethanyzhang Good point! Yes, we need both release notes and documentation.

steveburnett commented:
I'll be happy to review when documentation is added to the PR.

xin-zhang2 commented:
Added the documentation and the release note.

@majetideepak I added the documentation in the  Presto C++ chapter rather than the Connectors chapter, as it is presto C++ specific. While these metrics are defined in velox, I was thinking if it is proper to list these metrics in Presto documentation. What's your thought?

@steveburnett Can you help review the documentation? Thanks.

cc @ethanyzhang

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/features.rst:
```diff
@@ -226,4 +226,30 @@ Users can enable collection of worker level metrics by setting the property:
\ No newline at end of file
+  metrics collection in Prometheus Data Format refer `here <https://github.com/prestodb/presto/tree/master/presto-native-execution#build-prestissimo>`_.
+
+  When Presto C++ workers communicate with S3 Filesystem with this property enabled, the following metrics are additionally collected.
```
```suggestion
  When Presto C++ workers communicate with an S3 filesystem with this property enabled, the following metrics are collected.
```

steveburnett replied:
```suggestion
  When Presto C++ workers communicate with an S3 filesystem with this property enabled, the following metrics are collected.
```

steveburnett reviewed (changes requested):
Thanks for the doc! Minor rephrasing suggestion, looks good otherwise.

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/features.rst:
```diff
@@ -226,4 +226,30 @@ Users can enable collection of worker level metrics by setting the property:
+  ===================================== ===========================================================
+  ``velox.s3.active_connections``       The number of connections open for S3 read operations.
+
+  ``velox.s3.started_uploads``          The number of S3 upload calls that started.
```
```suggestion
  ``velox.s3.started_uploads``          The number of S3 upload calls that were started.
```

steveburnett commented on presto-docs/src/main/sphinx/presto_cpp/features.rst:
```diff
@@ -252,4 +252,7 @@ Users can enable collection of worker level metrics by setting the property:
+  metrics collection in Prometheus Data Format refer `here <https://github.com/prestodb/presto/tree/master/presto-native-execution#build-prestissimo>`_.
+
+  When enabled and Presto C++ workers interact with the S3 filesystem, additional runtime metrics are collected.
+  For a detailed list of these metrics, see `runtime metrics <https://facebookincubator.github.io/velox/monitoring/metrics.html#s3-filesystem>`_.
```
```suggestion
  For a detailed list of these metrics, see `S3 FileSystem <https://facebookincubator.github.io/velox/monitoring/metrics.html#s3-filesystem>`_.
```
Suggest using the text of the heading that is linked to, to reduce confusion in the reader when they look for "runtime metrics" which is at the top of this long page.

steveburnett reviewed (changes requested):
Thanks for the doc change! I think this is a good solution. Couple of suggestions about the links.

steveburnett reviewed (approved):
LGTM! (docs)

Pull updated branch, new local doc build, tested links. Everything looks good. Thanks!

xin-zhang2 commented:
@steveburnett  Thank your for the review!

@majetideepak  Could you take a look?

majetideepak commented:
Sorry! I missed the fact that the commits were not squashed.
